packages/@react-aria/dnd/src/useDroppableCollection.ts

```diff
@@ -49,6 +49,69 @@ export function useDroppableCollection(props: DroppableCollectionOptions, state:
   localState.props = props;
   localState.state = state;
 
+  let defaultOnDrop = useCallback(async (e: DroppableCollectionDropEvent) => {
+    let {
+      onInsert,
+      onRootDrop,
+      onItemDrop,
+      onReorder,
+      getDropOperation,
+      acceptedDragTypes = 'all'
+      // isValidDropTarget
+    } = localState.props;
+
+    let {draggingCollectionRef, draggingKeys} = getDnDState();
+    let isInternalDrop = draggingCollectionRef?.current === ref?.current;
+    let {
+      target,
+      dropOperation,
+      items
+    } = e;
+
+    let filteredItems = items;
+    // TODO: for now users who specify their own getDropOperation but use our utility drop handlers will have to perform valid drop item filtering in the utility drophandlers
+    // maybe change the expectation so that isValidDropTarget and acceptedDragTypes isn't just for the defaultGetDropOperation but should be provided if you are going to be using the util handlers?
+    if (!getDropOperation) {
+      filteredItems = items.filter((item) => {
+        let itemTypes = [];
+        if (item.kind === 'directory') {
+          return acceptedDragTypes === 'all' || acceptedDragTypes.includes('directory');
+        } else {
+          itemTypes = item.kind === 'file' ? [item.type] : [...item.types];
+        }
+
+        // TODO: fix this logic, maybe adjust what isValidDropTarget accepts. We need to in case the dropped on folder has extra restriction compared to the collection itself.
+        // check to see if there is a way to convert back to dragTypes. Maybe add another prop call? Or add another arg to isValidDropTarget for item which would only be defined during onDrop but not
+        // getDropOperation
+        return acceptedDragTypes === 'all' || itemTypes.some(type => acceptedDragTypes.includes(type));
+        // if (validDragType) {
+        //   // return (target.type === 'root' || target.dropPosition !== 'on' || (target.dropPosition === 'on' && isValidDropTarget(target, itemTypes)));
+        // }
+      });
```

Comment:
Upon further discussion we've decided to do some filtering for the user in onDrop based on `acceptedDragTypes` and `isValidDropTarget`. Still musing about what to do if the user has provided `getDropOperations` themselves but is still using our utility handlers OR how to check what items are actually valid for a onItemDrop since the folder/item might have additional type restrictions compared to the collection's acceptedDragTypes.

Perhaps we could make `acceptedDragTypes` a prop that the user provides even if `getDropOperations` is provided? Or maybe we can reuse `getDropOperations` and `isValidDropTarget`/`shouldAcceptItemDrop` and pass the DropItem as an extra parameter? It might be weird since the DropItem would be undefined until the drop actually happens. `getDropOperations` + `isValidDropTarget`/`shouldAcceptItemDrop` would have potentially two parts/additional complexity in their logic that the user would need to implement since the functions would get two kinds of information due to what info is available during `dragover` (aka we don't have the drag item's actual type/kind) and what is available during `drop` (we know the item's kind/types during this time).

This relates somewhat to the "Investigate getDropOperation `types` return consistency" issue since we could resolve this if we make getDropOperation's `types` general enough that it can take DragTypes and/or an array of the item's actual types. Alternatively we could add another prop that the user should provide just for filtering the accepted dropped items when dropping on a folder/item

Reply:
I think we should treat these two things separately:

* If you provide `getDropOperations`, it overrides `acceptedDragTypes` when handling dragover etc. i.e. determining whether to accept the drop.
* If both `acceptedDragTypes` and `getDropOperations` are provided, `acceptedDragTypes` would still be used in the default onDrop handler to do filtering. If you don't want that, then don't provide `acceptedDragTypes` and do the filtering yourself in onDrop.

Reply:
That sounds fine to me, I'll change acceptedDragTypes so that it doesn't default to `all` then. Any thoughts on the `types` we provide to `isValidDropTarget`/`shouldAcceptItemDrop`? I want to make them consistent when provided to the user in the default `getDropOperation` and when provided in the default `onDrop` handler, ideally it would always be of type `DragTypes`. To do so I was thinking of trying to pass the `e.dataTransfer` over to `option.onDrop` in useDrop so that I could construct `DragTypes`

Reply:
hmm. well `DragTypes` is only needed because there is some ambiguity on the types during `getDropOperation`. But in `onDrop` there is no ambiguity - we know all of the types. So we could, but it wouldn't really be necessary. Why do you think they need to be the same?

Reply:
hmm I think we should define `shouldAcceptItemDrop` to use `DragTypes` the same way `getDropOperation` does. But in the Typescript that's defined this way:

```ts
interface DragTypes {
  has(type: string): boolean
}
```

The exact class we pass shouldn't matter as long as it follows that interface. A `Set<string>` should conform to that type since it has a `has` method. So I think it's ok to pass the instance of our special class during `getDropOperation`, and a `Set` during `shouldAcceptItemType`.